File: test/javascript/erlang/persistent_term_test.mjs
"use strict";

import {
  assert,
  defineGlobalErlangAndElixirModules,
} from "../support/helpers.mjs";

import Erlang_Persistent_Term from "../../../assets/js/erlang/persistent_term.mjs";
import MemoryStorage from "../../../assets/js/memory_storage.mjs";
import Type from "../../../assets/js/type.mjs";

defineGlobalErlangAndElixirModules();

// IMPORTANT!
// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/erlang/persistent_term_test.exs
// Always update both together.

describe("Erlang_Persistent_Term", () => {
  beforeEach(() => {
    MemoryStorage.data = {};
  });

  describe("get/2", () => {
    const get = Erlang_Persistent_Term["get/2"];

    it("key exists", () => {
      const encodedKey = "tuple(atom(persistent_term),atom(my_key))";
      const value = Type.integer(123);
      MemoryStorage.put(encodedKey, value);

      const result = get(Type.atom("my_key"), Type.integer(234));

      assert.deepStrictEqual(result, value);
    });

    it("key doesn't exist", () => {
      const defaultValue = Type.integer(234);

      const result = get(Type.atom("my_key"), defaultValue);

      assert.deepStrictEqual(result, defaultValue);
    });
  });
});

File: test/elixir/hologram/ex_js_consistency/erlang/persistent_term_test.exs
defmodule Hologram.ExJsConsistency.Erlang.PersistentTermTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/erlang/persistent_term_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true
  alias Hologram.ExJsConsistency.Erlang.PersistentTermTest

  @moduletag :consistency

  describe "get/2" do
    test "key exists" do
      key = {PersistentTermTest, :my_key_1}
      :persistent_term.put(key, 123)

      assert :persistent_term.get(key, 234) == 123
    end

    test "key doesn't exist" do
      key = {PersistentTermTest, :my_key_2}
      assert :persistent_term.get(key, 234) == 234
    end
  end
end
